ObjectDisposedException from System.Net.Mail tests on Mono in CI

https://helix.dot.net/api/2019-06-17/jobs/4889c660-1974-4de2-8c00-cf0917b490fa/workitems/System.Net.Mail.Functional.Tests/console
```
  Discovering: System.Net.Mail.Functional.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Mail.Functional.Tests (found 144 of 146 test cases)
  Starting:    System.Net.Mail.Functional.Tests (parallel test collections = on, max threads = 2)

Unhandled Exception:
System.ObjectDisposedException: Safe handle has been closed.
Object name: 'SafeHandle'.
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.Net.Sockets.SocketPal.SetLingerOption(SafeSocketHandle handle, LingerOption optionValue)
   at System.Net.Sockets.Socket.SetLingerOption(LingerOption lref)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Object optionValue)
   at System.Net.Sockets.Socket.set_LingerState(LingerOption value)
   at System.Net.Sockets.TcpClient.set_LingerState(LingerOption value)
   at System.Net.Mail.SmtpConnection.Abort()
   at System.Net.Mail.SmtpTransport.Abort()
   at System.Net.Mail.SmtpClient.Abort()
   at System.Net.Mail.SmtpClient.TimeOutCallback(Object state)
   at System.Threading.TimerQueueTimer.<>c.<.cctor>b__23_0(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool)
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
   at System.Threading.TimerQueue.FireNextTimers()
   at System.Threading.TimerQueue.System.Threading.IThreadPoolWorkItem.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
[ERROR] FATAL UNHANDLED EXCEPTION: System.ObjectDisposedException: Safe handle has been closed.
Object name: 'SafeHandle'.
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.Net.Sockets.SocketPal.SetLingerOption(SafeSocketHandle handle, LingerOption optionValue)
   at System.Net.Sockets.Socket.SetLingerOption(LingerOption lref)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Object optionValue)
   at System.Net.Sockets.Socket.set_LingerState(LingerOption value)
   at System.Net.Sockets.TcpClient.set_LingerState(LingerOption value)
   at System.Net.Mail.SmtpConnection.Abort()
   at System.Net.Mail.SmtpTransport.Abort()
   at System.Net.Mail.SmtpClient.Abort()
   at System.Net.Mail.SmtpClient.TimeOutCallback(Object state)
   at System.Threading.TimerQueueTimer.<>c.<.cctor>b__23_0(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool)
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
   at System.Threading.TimerQueue.FireNextTimers()
   at System.Threading.TimerQueue.System.Threading.IThreadPoolWorkItem.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```
cc: @akoeplinger

--- a/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs
+++ b/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs
@@ -1078,18 +1078,7 @@ namespace System.Net.Mail
             }
         }
 
-        private void Abort()
-        {
-            try
-            {
-                _transport.Abort();
-            }
-            catch (ObjectDisposedException)
-            {
-                // See https://github.com/dotnet/corefx/issues/40711, and potentially
-                // catch additional exception types here if need demonstrates.
-            }
-        }
+        private void Abort() => _transport.Abort();
 
         public void Dispose()
         {

--- a/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs
+++ b/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs
@@ -1080,16 +1080,15 @@ namespace System.Net.Mail
 
         private void Abort()
         {
-            // Try-catch commented out to identify exception that is causing
-            // https://github.com/dotnet/corefx/issues/40711.
-
-            // try
-            // {
+            try
+            {
                 _transport.Abort();
-            // }
-            // catch
-            // {
-            // }
+            }
+            catch (ObjectDisposedException)
+            {
+                // See https://github.com/dotnet/corefx/issues/40711, and potentially
+                // catch additional exception types here if need demonstrates.
+            }
         }
 
         public void Dispose()

--- a/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpConnection.cs
+++ b/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpConnection.cs
@@ -153,25 +153,30 @@ namespace System.Net.Mail
         {
             if (!_isClosed)
             {
-                lock (this)
+                try
                 {
-                    if (!_isClosed && _tcpClient != null)
+                    lock (this)
                     {
-                        //free CBT buffer
-                        if (_channelBindingToken != null)
+                        if (!_isClosed && _tcpClient != null)
                         {
-                            _channelBindingToken.Close();
-                        }
+                            // free CBT buffer
+                            _channelBindingToken?.Close();
 
-                        // must destroy manually since sending a QUIT here might not be
-                        // interpreted correctly by the server if it's in the middle of a
-                        // DATA command or some similar situation.  This may send a RST
-                        // but this is ok in this situation.  Do not reuse this connection
-                        _tcpClient.LingerState = new LingerOption(true, 0);
-                        _networkStream?.Close();
-                        _tcpClient.Dispose();
+                            // must destroy manually since sending a QUIT here might not be
+                            // interpreted correctly by the server if it's in the middle of a
+                            // DATA command or some similar situation.  This may send a RST
+                            // but this is ok in this situation.  Do not reuse this connection
+                            _tcpClient.LingerState = new LingerOption(true, 0);
+                            _networkStream?.Close();
+                            _tcpClient.Dispose();
+                        }
+                        _isClosed = true;
                     }
-                    _isClosed = true;
+                }
+                catch (ObjectDisposedException)
+                {
+                    // See https://github.com/dotnet/corefx/issues/40711, and potentially
+                    // catch additional exception types here if need demonstrates.
                 }
             }
             _isConnected = false;

--- a/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpConnection.cs
+++ b/src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpConnection.cs
@@ -159,10 +159,9 @@ namespace System.Net.Mail
                     {
                         if (!_isClosed && _tcpClient != null)
                         {
-                            // free CBT buffer
                             _channelBindingToken?.Close();
 
-                            // must destroy manually since sending a QUIT here might not be
+                            // Must destroy manually since sending a QUIT here might not be
                             // interpreted correctly by the server if it's in the middle of a
                             // DATA command or some similar situation.  This may send a RST
                             // but this is ok in this situation.  Do not reuse this connection

